text/3750-cfg-os-version-min.md

```diff
@@ -0,0 +1,194 @@
+    os_version_min("tvos", "14.0"),
+    os_version_min("watchos", "7.0"),
+    os_version_min("visionos", "1.0")
+)]
+let preadv = {
+    extern "C" {
+        fn preadv(libc::c_int, *const libc::iovec, libc::c_int, off64_t) -> isize;
+    }
+    Some(preadv)
+};
+
+// Otherwise `preadv` needs to be weakly linked.
+// We do that using a `weak!` macro, defined elsewhere.
+#[cfg(not(any(
+    os_version_min("macos", "11.0"),
+    os_version_min("ios", "14.0"),
+    os_version_min("tvos", "14.0"),
+    os_version_min("watchos", "7.0"),
+    os_version_min("visionos", "1.0")
+)))]
+weak!(fn preadv(libc::c_int, *const libc::iovec, libc::c_int, off64_t) -> isize);
+
+if let Some(preadv) = preadv {
+    preadv(...) // Use preadv, it's available
+} else {
+    // ... fallback impl
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The `os_version_min` predicate allows users to conditionally compile code based on the API version supported by the target platform using `cfg`.
+It requires a key and a version string.
+The key can be either a `target_os` string or else one of a set of target-defined strings.
+Version strings are always target defined (see [Versioning Schema][versioning-schema]) and will be compared against the target's supported version.
+For example, `#[cfg("macos", "11.0")]` has the key `macos` and the minimum version `11.0`, which will match any macos version greater than or equal to `11.0`.
+If a target doesn't support a key, then the `cfg` will always return `false`.
+
+Each target platform will set the minimum API versions it supports for each key.
+
+## The standard library
+[the-standard-library]: #the-standard-library
+
+Currently the standard library is pre-compiled meaning that only a single version of each API can be supported, which must be the minimum version.
+Third party crates can choose to use a higher API level so long as it's compatible with the baseline API version.
+However, there is currently no support for setting a `os_version_min` version above the target's baseline (see [Future Possibilities][future-possibilities]).
+
+## Versioning Schema
+[versioning-schema]: #versioning-schema
+
+Version strings can take on nearly any form and while there are some standard formats, such as semantic versioning or release dates, projects can change schemas or provide aliases for some or all of their releases.
+Because of this diversity in version strings each platform will be responsible for defining an `is_gte` comparison function for each key they support.
+Alternatively they can use one of the pre-defined functions.
+
+## Linting
+[linting]: #linting
+
+By default `os_version_min` will be linted by `check_cfg` in a similar way to `target_os`.
+That is, all valid values for `target_os` will be accepted as valid keys for `os_version_min` on all platforms.
+The list of additional keys supported by the target will be consulted, which will then be allowed on a per-target basis.
```

Comment:
Remark: I want to note here that I want version validation, and I want it everywhere ;).

E.g. writing just `#[cfg(os_version_min("macos", "invalid"))]` should give an error stating that `invalid` is not a valid version string, and ideally regardless of what target I am currently compiling for.

Reply:
That might work for built-in targets but it can't work for external targets when we don't have the target spec (which admittedly are still unstable but still).

Reply:
I guess I'm doubtful of the benefit of this feature to external targets, but yeah, I'd be fine with not having validation for those.